File: public/code/reversal__counterattack.ts
import { Candle, Signal, makeSignal, atr, rsi } from "../indicators";

const body = (c: Candle) => Math.abs(c.close - c.open);
const rng = (c: Candle) => c.high - c.low;
const green = (c: Candle) => c.close > c.open;
const red = (c: Candle) => c.close < c.open;
const mkR = (c: Candle[], i: number, side: "long" | "short", slPrice: number, conf: number, reason: string): Signal => {
  const cur = c[i];
  if (side === "long") { const r = cur.close - slPrice; if (r <= 0) return makeSignal({ reason: "Invalid risk" }); return makeSignal({ signal: "long", entry: cur.close, stop_loss: slPrice, take_profit: [cur.close + r * 1.5, cur.close + r * 2.5, cur.close + r * 4], confidence: conf, reason }); }
  const r = slPrice - cur.close; if (r <= 0) return makeSignal({ reason: "Invalid risk" }); return makeSignal({ signal: "short", entry: cur.close, stop_loss: slPrice, take_profit: [cur.close - r * 1.5, cur.close - r * 2.5, cur.close - r * 4], confidence: conf, reason });
};

export function counterattack(c: Candle[]): Signal {
  if (c.length < 30) return makeSignal({ reason: "Insufficient data" });
  const i = c.length - 1, a = atr(c, 14), prev = c[i - 1], cur = c[i];
  // Bullish: büyük kırmızı + yeşil aynı close'a kapanır (counterattack)
  if (red(prev) && body(prev) > a[i] && green(cur) && Math.abs(cur.close - prev.close) / prev.close < 0.004 && cur.open < prev.close) return mkR(c, i, "long", cur.low - 0.4 * a[i], 0.67, "Bullish counterattack line");
  if (green(prev) && body(prev) > a[i] && red(cur) && Math.abs(cur.close - prev.close) / prev.close < 0.004 && cur.open > prev.close) return mkR(c, i, "short", cur.high + 0.4 * a[i], 0.67, "Bearish counterattack line");
  return makeSignal({ reason: "No counterattack" });
}

File: public/code/reversal__counterattack.py
from common import make_signal, ema, sma, rsi, macd, atr, bollinger_bands, vwap, swing_highs, swing_lows
import math


def _sign(x):
    return (x > 0) - (x < 0)


body = lambda c: abs(c['close'] - c['open'])
rng = lambda c: c['high'] - c['low']
green = lambda c: c['close'] > c['open']
red = lambda c: c['close'] < c['open']
def mkR(c, i, side, slPrice, conf, reason):
    cur = c[i]
    if side == "long":
        r = cur['close'] - slPrice
        if r <= 0:
            return make_signal(reason="Invalid risk")
        return make_signal(signal="long", entry=cur['close'], stop_loss=slPrice, take_profit=[cur['close'] + r * 1.5, cur['close'] + r * 2.5, cur['close'] + r * 4], confidence=conf, reason=reason)
    r = slPrice - cur['close']
    if r <= 0:
        return make_signal(reason="Invalid risk")
    return make_signal(signal="short", entry=cur['close'], stop_loss=slPrice, take_profit=[cur['close'] - r * 1.5, cur['close'] - r * 2.5, cur['close'] - r * 4], confidence=conf, reason=reason)
def counterattack(c):
    if len(c) < 30:
        return make_signal(reason="Insufficient data")
    i = len(c) - 1
    a = atr(c, 14)
    prev = c[i - 1]
    cur = c[i]
    # Bullish: büyük kırmızı + yeşil aynı close'a kapanır (counterattack)
    if red(prev)  and  body(prev) > a[i]  and  green(cur)  and  abs(cur['close'] - prev['close']) / prev['close'] < 0.004  and  cur['open'] < prev['close']:
        return mkR(c, i, "long", cur['low'] - 0.4 * a[i], 0.67, "Bullish counterattack line")
    if green(prev)  and  body(prev) > a[i]  and  red(cur)  and  abs(cur['close'] - prev['close']) / prev['close'] < 0.004  and  cur['open'] > prev['close']:
        return mkR(c, i, "short", cur['high'] + 0.4 * a[i], 0.67, "Bearish counterattack line")
    return make_signal(reason="No counterattack")
